resdk/data_upload/annotate_samples.py

```diff
@@ -0,0 +1,85 @@
+"""Annotate samples in a collection."""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+from resdk.data_upload.samplesheet import FileImporter, Sample
+
+def annotate_samples(collection, samplesheet_name, local_sheet=True,
+                     verbose=True):
+    """Annotate the samples in a collection based on a sample annotation
+    spreadsheet.
+    """
+
+    # Extract the information from a sample annotation spreadsheet
+    # Validate each sample annotation
+    if local_sheet:
+        samplesheet = FileImporter(samplesheet_name)
+        err_samples = []
+        missing_samples = []
+        sample_count = 0
+
+    # TODO: Try to pull the samplesheet as a data object from the collection
+    #       once it is possible to upload samplesheets
+    # else:
+    #     sheet = collection.data.get(name=samplesheet_name,
+    #                                 type='data:annotation')
+    #     sheet.download()
+    #     samplesheet = FileImporter(sheet.files()[0])
+    #     err_samples = []
+
+    # Apply each annotation to a sample in the collection
+    for entry in samplesheet.sample_list:
+        try:
+            # Find the existing (unannotated) samples
+            new_samples = collection.samples.filter(name=entry['SAMPLE_NAME'])
+            if not new_samples:
+                raise NameError(
+                    "Sample `{}` not found.".format(entry['SAMPLE_NAME'])
+                )
+
+            # Create and validate the sample descriptor
+            sample_info = Sample(entry)
```

Comment:
Move this to `FileImporter`, so it returns a list of `Sample` objects.

Reply:
I did it this way to give the NameError associated with missing samples a higher priority than the ValueError associated with invalid data. If the list of Sample objects is created first, then they must all be validated first (before being compared to the collection samples), and the ValueErrors will necessarily trigger before the NameErrors. 

One way around this would be to have `FileImporter` keep track of invalid samples without eliminating them altogether, and then let `annotate_samples` filter them out later, after checking for missing samples. The advantage here is that then both missing samples and invalid data could be reported for the same entries. I originally thought this would be a needlessly convoluted approach, but do you think it may be worth it?